File: src/ImageExtractor.java

import java.awt.image.BufferedImage;
import java.io.IOException;
import java.util.LinkedList;
import java.util.Stack;

public class ImageExtractor extends Main {
  private static class Coords {
    private final int x, y;

    public Coords(int x, int y) {
      this.x = x;
      this.y = y;
    }
  }
  
  private static final Stack<Coords> coordsStack = new Stack<>();
  
  public static final LinkedList<LinkedList<Image>> images = new LinkedList<>();
  
  public static void process(boolean[] screen, boolean[] background, int frame
      , BufferedImage backgroundImage) throws IOException {
    final int SAME = 0, CHANGED = 1;
    int x1, y1, x2, y2, imageNumber = 1;
    int[] pixels = new int[PIXEL_SIZE];
    
    int changed = 0;
    for(int addr = 0; addr < PIXEL_SIZE; addr++) {
      boolean isChanged = screen[addr] != background[addr];
      pixels[addr] = isChanged ? CHANGED : SAME;
      if(isChanged) changed++;
    }
    
    if(changed > MAX_CHANGED_PIXELS) {
      System.out.println("  Changed pixels of " + frame + " is " + changed);
      return;
    }
    
    for(int y = 0; y < PIXEL_HEIGHT; y++) {
      int ySource = y << 8;
      x: for(int x = 0; x < PIXEL_WIDTH; x++) {
        int addr = ySource | x;
        if(pixels[addr] == CHANGED) {
          x1 = x2 = x;
          y1 = y2 = y;
          imageNumber++;
          pixels[addr] = imageNumber;
          coordsStack.add(new Coords(x, y));
          while(!coordsStack.empty()) {
            Coords coords = coordsStack.pop();
            int x0 = coords.x;
            int y0 = coords.y;
            for(int dy = -MAX_DISTANCE; dy <= MAX_DISTANCE; dy++) {
              int yy = y0 + dy;
              int yAddr = yy * PIXEL_WIDTH;
              if(yy < 0 || yy >= PIXEL_HEIGHT) continue;
              for(int dx = -MAX_DISTANCE; dx <= MAX_DISTANCE; dx++) {
                if(dx == 0 && dy == 0) continue;
                int xx = x0 + dx;
                if(xx < 0 || xx >= PIXEL_WIDTH) continue;
                int addr2 = x0 + dx + yAddr;
                if(pixels[addr2] == CHANGED) {
                  x1 = Integer.min(x1, xx);
                  y1 = Integer.min(y1, yy);
                  x2 = Integer.max(x2, xx);
                  y2 = Integer.max(y2, yy);
                  pixels[addr2] = imageNumber;
                  coordsStack.add(new Coords(xx, yy));
                }
              }
            }
          } 
          x2++;
          y2++;
          
          if(Image.hasAcceptableSize(x1, y1, x2, y2)) {
            if(x2 - x1 >= MAX_WIDTH || y2 - y1 >= MAX_HEIGHT) {
              System.out.println("  " + frame + " - skipped big "
                  + (x2 - x1) + "x" + (y2 - y1) + "changed area.");
            }
            
            if(mode == Mode.DETECT_MAX_SIZE) continue;
            if(mode == Mode.EXTRACT_SPRITES) {
              Image image = new Image(pixels, screen, background, x1, y1, x2, y2
                  , imageNumber);
              for(LinkedList<Image> list: images) {
                for(Image listImage: list) {
                  switch(listImage.compareTo(image)) {
                    case EQUAL:
                      continue x;
                    case SIMILAR:
                      list.add(image);
                      continue x;
                    case OTHER:
                      break;
                  }
                }
              }

              LinkedList<Image> newList = new LinkedList<>();
              newList.add(image);
              images.add(newList);
            } else {
              Sprites.declash(pixels, imageNumber, screen, background
                  , x1, y1, x2, y2, backgroundImage);
            }
          } else {
            for(int yy = y1; yy < y2; yy++) {
              int yAddr = yy * PIXEL_WIDTH;
              for(int xx = x1; xx < x2; xx++) {
                if(pixels[xx + yAddr] == imageNumber) {
                  backgroundImage.setRGB(xx, yy, particleColor);
                }
              }
            }
          }
        }
      }
    }
    Screen.saveImage(backgroundImage, frame);
  }

  public static void saveImages() throws IOException {
    int listNum = 0;
    for(LinkedList<Image> list: images) {
      int maxWeight = -1;
      Image maxImage = null;
      for(Image image: list) {
        int size = image.getWeight();
        if(maxWeight < size) {
          maxWeight = size;
          maxImage = image;
        }
      }
      listNum++;
      maxImage.save(listNum);
    }
  }
}

File: src/ImageExtractor.kt
import Image.Comparsion
import java.awt.Color
import java.awt.image.BufferedImage
import java.io.IOException
import java.util.*

object ImageExtractor {
  private val coordsStack = Stack<Coords>()
  val images = LinkedList<LinkedList<Image>>()

  @Throws(IOException::class)
  fun process(screen: Area, background: Background?, frame: Int
              , image: BufferedImage) {
    if(background == null) return

    val SAME = 0
    val CHANGED = 1
    val pixels = IntArray(PIXEL_SIZE)
    var changed = 0
    val backgroundValues = background.pixels
    for(addr in 0 until PIXEL_SIZE) {
      val pixel = screen.pixels[addr]
      val bgPixel = backgroundValues[addr]
      var isChanged = pixel != bgPixel && (ANY_IS_CHANGED || bgPixel != Pixel.ANY)
      pixels[addr] = if(isChanged) CHANGED else SAME
      if(isChanged) changed++
    }

    /*if(changed > MAX_CHANGED_PIXELS) {
      println("  Changed pixels of $frame is $changed")
      return
    }*/

    val areas = LinkedList<ChangedArea>()
    var imageNumber = 1
    for(y in 0 until PIXEL_HEIGHT) {
      val ySource = y shl 8
      x@ for(x in 0 until PIXEL_WIDTH) {
        val addr = ySource or x
        if(pixels[addr] == CHANGED) {
          var x2: Int = x
          var x1: Int = x2
          var y2: Int = y
          var y1: Int = y2
          imageNumber++
          pixels[addr] = imageNumber
          coordsStack.add(Coords(x, y))
          while(!coordsStack.empty()) {
            val coords = coordsStack.pop()
            val x0 = coords.x
            val y0 = coords.y
            for(dy in -MAX_DISTANCE..MAX_DISTANCE) {
              val yy = y0 + dy
              if(yy < 0 || yy >= PIXEL_HEIGHT) continue
              val yAddr = yy * PIXEL_WIDTH
              for(dx in -MAX_DISTANCE..MAX_DISTANCE) {
                if(dx == 0 && dy == 0) continue
                val xx = x0 + dx
                if(xx < 0 || xx >= PIXEL_WIDTH) continue
                val addr2 = x0 + dx + yAddr
                if(pixels[addr2] == CHANGED) {
                  x1 = Integer.min(x1, xx)
                  y1 = Integer.min(y1, yy)
                  x2 = Integer.max(x2, xx)
                  y2 = Integer.max(y2, yy)
                  pixels[addr2] = imageNumber
                  coordsStack.add(Coords(xx, yy))
                }
              }
            }
          }

          x2++
          y2++
          if(Image.Companion.hasAcceptableSize(x1, y1, x2, y2)) {
            if(x2 - x1 >= MAX_WIDTH || y2 - y1 >= MAX_HEIGHT) {
              println("  $frame - skipped big ${x2 - x1}" +
                  " x ${y2 - y1} changed area.")
              continue
            }
            areas.add(ChangedArea(imageNumber, x1, y1, x2, y2))
            if(mode === Mode.DETECT_MAX_SIZE) continue
            if(mode === Mode.EXTRACT_SPRITES) {
              val image = Image(pixels, screen.pixels, backgroundValues
                , x1, y1, x2, y2, imageNumber)
              for(list in images) {
                for(listImage in list) {
                  when(listImage.compareTo(image)) {
                    Comparsion.EQUAL -> {
                      listImage.incrementQuantity()
                      continue@x
                    }
                    Comparsion.SIMILAR -> {
                      list.add(image)
                      continue@x
                    }
                  }
                }
              }
              val newList = LinkedList<Image>()
              newList.add(image)
              images.add(newList)
            } else {
              repaint(pixels, imageNumber, screen, x1, y1, x2, y2, image
                , background.particlesArea)
            }
          } else {
            repaint(pixels, imageNumber, screen, x1, y1, x2, y2, image
              , background.particlesArea)
          }
        }
      }
    }

    if(mode == Mode.DECLASH)
      Sprites.declash(screen, image, background.frame, areas)
  }

  private fun repaint(pixels: IntArray, imageNumber: Int, screen: Area
                      , x1: Int, y1: Int, x2: Int, y2: Int
                      , image: BufferedImage, particlesArea: Rect?) {
    if(SHOW_DETECTION_AREA && particlesArea != null) {
      particlesArea.draw(image)
    }
    for(y in y1 until y2) {
      val yAddr = y * PIXEL_WIDTH
      val yAttrSource = (y shr 3) shl 5
      for(x in x1 until x2) {
        val addr = x + yAddr
        if(pixels[addr] == imageNumber) {
          val attr = screen.attrs[yAttrSource or (x shr 3)]
          image.setRGB(x, y, if(screen.pixels[addr] == Pixel.ANY) {
            if(SHOW_DETECTION_AREA) magenta else color[attr and 0xF]
          } else if(particlesArea != null && particlesArea.has(x, y)) {
            PARTICLE_COLOR
          } else if(screen.pixels[addr] == Pixel.ON) {
            color[attr and 0xF]
          } else {
            color[attr shr 4]
          })
        } else if(SHOW_DETECTION_AREA) {
          image.setRGB(x, y, darkMagenta)
        }
      }
    }
  }

  @Throws(IOException::class)
  fun saveImages() {
    var listNum = 0
    for(list in images) {
      var maxWeight = -1
      var maxImage: Image = list.first
      for(image in list) {
        val size = image.weight
        if(maxWeight < size) {
          maxWeight = size
          maxImage = image
        }
      }
      listNum++
      maxImage.save(listNum)
    }
  }
}